src/Bicep.Core/Emit/EmitLimitationCalculator.cs

```diff
@@ -446,5 +448,31 @@ private static bool IsInvariant(SemanticModel semanticModel, LocalVariableSyntax
                 ? IsLocalInvariant(itemVariable)
                 : IsLocalInvariant(itemVariable) && IsLocalInvariant(indexVariable);
         }
+
+        private static void BlockModuleOutputResourcePropertyAccess(SemanticModel model, IDiagnosticWriter diagnostics) =>
+            diagnostics.WriteMultiple(model.AllResources.OfType<ModuleOutputResourceMetadata>()
+                // find references to this resource
+                .SelectMany(moduleOutputResource => model.FindReferences(moduleOutputResource.Module)
+                    .Select(moduleReference => GetMeaningfulParent(model, moduleReference))
+                    .OfType<PropertyAccessSyntax>()
+                    .Where(pas => LanguageConstants.IdentifierComparer.Equals(pas.PropertyName.IdentifierName, "outputs"))
+                    .Select(outputsReference => GetMeaningfulParent(model, outputsReference))
+                    .OfType<PropertyAccessSyntax>()
+                    .Where(pas => LanguageConstants.IdentifierComparer.Equals(pas.PropertyName.IdentifierName, moduleOutputResource.OutputName)))
+                // of the found references, find any that attempt to access the resource's properties
+                .Where(moduleOutputResourceAccessSyntax => GetMeaningfulParent(model, moduleOutputResourceAccessSyntax) is PropertyAccessSyntax resourcePropertyAccess &&
+                    !ResourcePropertiesResolvedAtCompileTime.Contains(resourcePropertyAccess.PropertyName.IdentifierName))
+                // generate an error diagnostic for each such access
+                .Select(syntaxToBlock => DiagnosticBuilder.ForPosition(syntaxToBlock).ModuleOutputResourcePropertyAccessDetected()));
+
+        private static ImmutableHashSet<string> ResourcePropertiesResolvedAtCompileTime
+            = ImmutableHashSet.CreateRange(LanguageConstants.IdentifierComparer, AzResourceTypeProvider.ReadWriteDeployTimeConstantPropertyNames);
```

Comment:
Should we just make `AzResourceTypeProvider.ReadWriteDeployTimeConstantPropertyNames` into an `ImmutableSortedSet`?